sdk/src/nonce_state.rs

```diff
@@ -0,0 +1,549 @@
+        rent: &Rent,
+    ) -> Result<(), InstructionError>;
+    fn set_nonce(
+        &mut self,
+        recent_blockhashes: &RecentBlockhashes,
+        meta: &Meta,
+    ) -> Result<(), InstructionError>;
+    fn withdraw(
+        &mut self,
+        lamports: u64,
+        to: &mut KeyedAccount,
+        recent_blockhashes: &RecentBlockhashes,
+        signers: &HashSet<Pubkey>,
+    ) -> Result<(), InstructionError>;
+}
+
+impl<'a> NonceAccount for KeyedAccount<'a> {
+    fn set_nonce(
+        &mut self,
+        recent_blockhashes: &RecentBlockhashes,
+        meta: &Meta,
+    ) -> Result<(), InstructionError> {
+        if recent_blockhashes.is_empty() {
+            Err(NonceError::NoRecentBlockhashes.into())
+        } else {
+            self.set_state(
+                &NonceState::Initialized(*meta, recent_blockhashes[0])
+            )
+        }
+    }
+
+    fn nonce(
+        &mut self,
+        recent_blockhashes: &RecentBlockhashes,
+        rent: &Rent,
+    ) -> Result<(), InstructionError> {
+        let meta = match self.state()? {
+            NonceState::Initialized(meta, ref hash) => {
+                if recent_blockhashes.contains(hash) {
+                    return Err(NonceError::NotReady.into());
+                }
+                meta
+            },
+            NonceState::Uninitialized => {
+                let min_balance = rent.minimum_balance(self.account.data.len());
+                if self.account.lamports < min_balance {
+                    return Err(InstructionError::InsufficientFunds.into());
+                }
+                Meta::new(min_balance)
+            },
+        };
+        self.set_nonce(recent_blockhashes, &meta)
+    }
+
+    fn withdraw(
+        &mut self,
+        lamports: u64,
+        to: &mut KeyedAccount,
+        recent_blockhashes: &RecentBlockhashes,
+        signers: &HashSet<Pubkey>,
+    ) -> Result<(), InstructionError> {
+        if !signers.contains(self.unsigned_key()) {
+            return Err(InstructionError::MissingRequiredSignature);
+        }
+
+        match self.state()? {
+            NonceState::Uninitialized => {
+                if lamports > self.account.lamports {
+                    return Err(InstructionError::InsufficientFunds.into());
+                }
+                self.set_state(&NonceState::Uninitialized)?;
+            },
+            NonceState::Initialized(meta, ref hash) => {
+                if lamports == self.account.lamports {
+                    if recent_blockhashes.contains(hash) {
+                        return Err(NonceError::NotReady.into());
+                    }
+                    self.set_state(&NonceState::Uninitialized)?;
+                } else if lamports + meta.rent_exempt_lamports > self.account.lamports {
+                    return Err(InstructionError::InsufficientFunds.into());
+                } else {
+                    self.set_nonce(recent_blockhashes, &meta)?;
```

Comment:
why Nonce() here?  can you add a comment for the "why"?

Reply:
It doesn't.  I had in my head for awhile that everything that modifies the account should be advancing the nonce.  Proposal convo shook that out, but changes hadn't made it here yet.